### DISCUSSION TITLE
Cannot connect to backend

### DISCUSSION THREAD

**Discussion by ilgigante77** [MDeLuise/plant-it#128] [Q&A] [answered]
I would like to use this app 'coz I've been searching for long time for a self-hosted open source gardening companio app, but I cannot manage how to make frontend and backend to communicate. I've tried the options of the installation guide, tryin with the docker host IP, hostname, internal docker backend container name, hostaname and IP, but I receive everytime a "cannot connect to..." message after 10 s... I've tried to get into the frontend container bash and try to curl, but I get a "connection refused" message... Maybe I missed something?

**Comment by MDeLuise** [comment]
Hi @ilgigante77, thank for opening the issue!
Could you please provide the content of the `docker-compose.yml`, the `backend.env` and the `frontend.env` files you're using?
Remember to change/remove sensitive information of them (like database password, trefle token, and so on)

**Comment by ilgigante77** [comment]
Here you are:

version: "3"
services:
  backend:
    image: msdeluise/plant-it-backend:latest
    env_file: backend.env
    depends_on:
      - db
      - cache
    restart: always
    volumes:
      - "/srv/plantit/uploads:/upload-dir"
      - "certs:/certificates"
    ports:
      - "5002:8080"

  db:
    image: mysql:8.0
    restart: always
    env_file: backend.env
    volumes:
      - "plantit-db:/var/lib/mysql"

  cache:
    image: redis:7.2.1
    restart: always

  frontend:
    image: msdeluise/plant-it-frontend:latest
    restart: always
    env_file: frontend.env
    links:
      - backend
    ports:
      - "5001:3000"
    volumes:
      - "certs:/certificates"
volumes:
  plantit-db:
  certs:
    driver: local
    driver_opts:
      type: none
      o: bind
      device: /srv/plantit/certificates

-----------------------------------------------------------

#
# DB
#
MYSQL_HOST=db
MYSQL_PORT=3306
MYSQL_USERNAME=root
MYSQL_PSW=xxxxxx
MYSQL_ROOT_PASSWORD=xxxxxx
MYSQL_DATABASE=bootdb

#
# JWT
#
JWT_SECRET=xxxxxxxxxxx
JWT_EXP=1

#
# Server config
#
USERS_LIMIT=-1
UPLOAD_DIR=/upload-dir
API_PORT=8080
TREFLE_KEY=
ALLOWED_ORIGINS=*
LOG_LEVEL=DEBUG
UPDATE_EXISTING=false

#
# Cache
#
CACHE_TTL=86400
CACHE_HOST=cache
CACHE_PORT=6379

#
# SSL
#
SSL_ENABLED=false
CERTIFICATE_PATH=/certificates/

-----------------------------------------------------------

PORT=3000
API_URL=http://<docker_host_ip>:5002/api
WAIT_TIMEOUT=10000
CACHE_TTL_DAYS=7
BROWSER=none
SSL_ENABLED=false
CERTIFICATE_PATH=/certificates/

-----------------------------------------------------------
The 5001 and 5002 ports are open on the firewall

**Reply by MDeLuise** [reply]
At a first look it seems to be all ok. 
* Is the `<docker_host_ip>` an internal ip like `192.168.x.x` or is it something else?
* Are you able to connect to `<docker_host_ip>:5002/api/swagger-ui/index.html`? It should be displayed the homepage of rest api

**Reply by ilgigante77** [reply]
Yes, the <docker_host_ip> is like 192.168.x.x and it's the LAN physical eth adapter address. 
No, the curl fails even if invoked by the server itself... 
If I do the lsof command from the docker host I get:
`docker-pr 3154506            root    4u  IPv4 45728820      0t0  TCP *:5002 (LISTEN)`
so it seems somebody is listening.
from inside the container, the app seems to be running:
`root          10  9.8  2.0 6479580 270508 ?      Sl   09:07   0:27 java -jar /opt/app/app.jar`
but every now and then the container itself seems to restart...

**Reply by MDeLuise** [reply]
> No, the curl fails even if invoked by the server itself...

This is strange, could you please provide the docker-compose log? I want to understand if the backend service have some problem and crashed.

**Reply by ilgigante77** [reply]
Are you talking about that?

Creating network "plantit_default" with the default driver
Pulling db (mysql:8.0)...
8.0: Pulling from library/mysql
81badc5f380f: Pull complete
6a81d7aa4708: Pull complete
a7d2a50df679: Pull complete
4b16e7be1767: Pull complete
3205e7b85ef5: Pull complete
d8b021d6b4e1: Pull complete
8df9a740b0d6: Pull complete
734b576d7682: Pull complete
c88d6cb0b217: Pull complete
20e828a3ad27: Pull complete
fb468cb0f95b: Pull complete
Digest: sha256:17326312dfb4eab40db4adeb94e853806c5b361caf80e5089bc37732077e2bec
Status: Downloaded newer image for mysql:8.0
Pulling cache (redis:7.2.1)...
7.2.1: Pulling from library/redis
a378f10b3218: Pull complete
3c75410c1f8b: Pull complete
667874757cc1: Pull complete
7150b93d249d: Pull complete
ed7c83735e28: Pull complete
4f4fb700ef54: Pull complete
3fa899a007ab: Pull complete
Digest: sha256:4ca2a277f1dc3ddd0da33a258096de9a1cf5b9d9bd96b27ee78763ee2248c28c
Status: Downloaded newer image for redis:7.2.1
Pulling backend (msdeluise/plant-it-backend:latest)...
latest: Pulling from msdeluise/plant-it-backend
d66d6a6a3687: Already exists
18f947fdc0fc: Pull complete
2087adbde65e: Pull complete
df0413a4cb90: Pull complete
f2a245455ba1: Pull complete
312e57a9c773: Pull complete
f87fe4c101dc: Pull complete
81a5ee8f4843: Pull complete
7a9191c5e412: Pull complete
b80ebace5125: Pull complete
f7fa4a71c00e: Pull complete
Digest: sha256:9921b041b1184f075ed2c92cdad98615765dd5baebd1353a9f762b503e96b5fa
Status: Downloaded newer image for msdeluise/plant-it-backend:latest
Pulling frontend (msdeluise/plant-it-frontend:latest)...
latest: Pulling from msdeluise/plant-it-frontend
911c6d0c7995: Pull complete
131e13eca73f: Pull complete
95376bf29516: Pull complete
6717402ec973: Pull complete
ec1346715d6e: Pull complete
e4a8165f783f: Pull complete
c8e5f923870c: Pull complete
e50ac71e9da9: Pull complete
4f4fb700ef54: Pull complete
606246bc4354: Pull complete
c9e0d3f32989: Pull complete
bd1a4378d15c: Pull complete
c6ac09b84040: Pull complete
2e10a40e215b: Pull complete
Digest: sha256:ff0f0adc18da58fe9d6edee5c422fdf3822798812aa7179e599a75722ff881da
Status: Downloaded newer image for msdeluise/plant-it-frontend:latest
Creating plantit_db_1 ...
Creating plantit_cache_1 ...
Pulling db (mysql:8.0)...
Pulling cache (redis:7.2.1)...
8.0: Pulling from library/mysql
7.2.1: Pulling from library/redis
a378f10b3218: Pulling fs layer
3c75410c1f8b: Pulling fs layer
667874757cc1: Pulling fs layer
7150b93d249d: Waiting
ed7c83735e28: Waiting
4f4fb700ef54: Waiting
3fa899a007ab: Waiting
81badc5f380f: Pull complete
6a81d7aa4708: Pull complete
a7d2a50df679: Pull complete
4b16e7be1767: Pull complete
3205e7b85ef5: Pull complete
d8b021d6b4e1: Pull complete
8df9a740b0d6: Extracting [>                                                  ]  557.1kB/58.52MB
7150b93d249d: Pull complete
8df9a740b0d6: Pull complete
734b576d7682: Pull complete
c88d6cb0b217: Pull complete
20e828a3ad27: Pull complete
fb468cb0f95b: Pull complete
Creating plantit_db_1    ... done
Status: Downloaded newer image for mysql:8.0
Creating plantit_backend_1 ... done
Creating plantit_frontend_1 ... done

**Reply by ilgigante77** [reply]
I understood now... Please see attachment
[dc.zip](https://github.com/MDeLuise/plant-it/files/14384032/dc.zip)
And looking at it, I found a missing dependency exception and a "cannot acquire lock" error.

**Reply by MDeLuise** [reply, marked as answer]
Great, I meant exactly that log.
Ok so it seems to be related to the connection between backend and db.
I saw that in you `docker-compose` you are using
```
db:
image: mysql:8.0
restart: always
env_file: backend.env
volumes:
  - "plantit-db:/var/lib/mysql"
```

Just for a quick test, could you try to use this?
```
db:
image: mysql:8.0
restart: always
env_file: backend.env
volumes:
  - "./db:/var/lib/mysql"
```

It should create a `db` directory inside the folder you're using

**Comment by ilgigante77** [comment]
eureka! now it works! I didn't know that the db was directly accessed by the frontend, so I created a virtual volume for it! thanks tho

**Reply by MDeLuise** [reply]
I will simplify the deployment in one of the next releases, I'm considering the removal of the `frontend` service in favor of a dedicated app.

Thank you, an happy planting! 🪴
